File: snow_pkg/src/openalea/Melting.py
# coding: utf8
from copy import copy

import numpy
from math import *

def model_Melting(jul = 0,
         Tmf = 0.5,
         DKmax = 0.0,
         Kmin = 0.0,
         tavg = 0.0):
    """
     - Name: Melting -Version: 1.0, -Time step: 1
     - Description:
                 * Title: snow in the process of melting
                 * Author: STICS
                 * Reference: -
                 * Institution: INRA
                 * Abstract: -
     - inputs:
                 * name: jul
                               ** min : 0
                               ** default : 0
                               ** max : 366
                               ** uri : 
                               ** variablecategory : auxiliary
                               ** datatype : INT
                               ** inputtype : variable
                               ** unit : d
                               ** description : current day of year for the calculation
                 * name: Tmf
                               ** parametercategory : constant
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 1.0
                               ** uri : 
                               ** default : 0.5
                               ** inputtype : parameter
                               ** unit : degC
                               ** description : threshold temperature for snow melting 
                 * name: DKmax
                               ** parametercategory : constant
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 5000.0
                               ** uri : 
                               ** default : 0.0
                               ** inputtype : parameter
                               ** unit : mmW/degC/d
                               ** description : difference between the maximum and the minimum melting rates
                 * name: Kmin
                               ** parametercategory : constant
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 5000.0
                               ** uri : 
                               ** default : 0.0
                               ** inputtype : parameter
                               ** unit : mmW/degC/d
                               ** description : minimum melting rate on 21 December
                 * name: tavg
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 100.0
                               ** uri : 
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : degC
                               ** description : average temperature
     - outputs:
                 * name: M
                               ** min : 0.0
                               ** variablecategory : rate
                               ** max : 500.0
                               ** uri : 
                               ** datatype : DOUBLE
                               ** unit : mmW/d
                               ** description : snow in the process of melting
    """

    M = 0.0
    K = DKmax / 2.0 * -sin((2.0 * pi * float(jul) / 366.0 + (9.0 / 16.0 * pi))) + Kmin + (DKmax / 2.0)
    if tavg > Tmf:
        M = K * (tavg - Tmf)
    return M

File: snow_pkg/src/dssat/Melting.f90
MODULE Meltingmod
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_Melting(jul, &
        Tmf, &
        DKmax, &
        Kmin, &
        tavg, &
        M)
        IMPLICIT NONE
        INTEGER, INTENT(IN) :: jul
        REAL, INTENT(IN) :: Tmf
        REAL, INTENT(IN) :: DKmax
        REAL, INTENT(IN) :: Kmin
        REAL, INTENT(IN) :: tavg
        REAL, INTENT(OUT) :: M
        REAL:: K
        !- Description:
    !            * Title: snow in the process of melting
    !            * Author: STICS
    !            * Reference: -
    !            * Institution: INRA
    !            * Abstract: -
        !- inputs:
    !            * name: jul
    !                          ** description : current day of year for the calculation
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : INT
    !                          ** default : 0
    !                          ** min : 0
    !                          ** max : 366
    !                          ** unit : d
    !                          ** uri : 
    !            * name: Tmf
    !                          ** description : threshold temperature for snow melting 
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** default : 0.5
    !                          ** min : 0.0
    !                          ** max : 1.0
    !                          ** unit : degC
    !                          ** uri : 
    !            * name: DKmax
    !                          ** description : difference between the maximum and the minimum melting rates
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 5000.0
    !                          ** unit : mmW/degC/d
    !                          ** uri : 
    !            * name: Kmin
    !                          ** description : minimum melting rate on 21 December
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 5000.0
    !                          ** unit : mmW/degC/d
    !                          ** uri : 
    !            * name: tavg
    !                          ** description : average temperature
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 100.0
    !                          ** unit : degC
    !                          ** uri : 
        !- outputs:
    !            * name: M
    !                          ** description : snow in the process of melting
    !                          ** variablecategory : rate
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 500.0
    !                          ** unit : mmW/d
    !                          ** uri : 
        M = 0.0
        K = DKmax / 2.0 * (-SIN((2.0 * 3.14159265 * REAL(jul) / 366.0 + (9.0  &
                / 16.0 * 3.14159265)))) + Kmin + (DKmax / 2.0)
        IF(tavg .GT. Tmf) THEN
            M = K * (tavg - Tmf)
        END IF
    END SUBROUTINE model_Melting

END MODULE
